### ISSUE TITLE
Distributed tasks

### ISSUE THREAD

**Issue by rymsha** [enonic/xp#8436]
Goals:
- it should be possible to submit a task that executes only once on an arbitrary node in the cluster
- it should be possible to specify which xp nodes can execute distributed tasks via configuration

Non-goals:
- there is no intention to implement distributed map-reduce algorithms. A submitted task executes exclusively on one node.
- there is no intention to make `lib-task.submit` method to benefit from this new feature. It accepts execution callback that can rely on existing variables in java script scope, not necessarily transferable to a remote node.

**Comment by rymsha** [comment]
There are 3 types of "tasks" in XP
- Management RunnableTasks in Java (`VacuumRunnableTask`, `ReindexRunnableTask`, etc)
- callback-like tasks (applications call `taskLib.summit` to execute them)
- named tasks (applications call `taskLib.submitNamed` to execute them)

First two are out of scope:
Management RunnableTasks in Java theoretically can be distributed, but they all depend on OSGi services. To make them to work across the cluster we have to make them `Serializable` (read - rewrite the code of each and every). It is way simpler to pin-point the node they should be run instead as they always start with an http request.

callback-like tasks are practically impossible to serialize as they can rely on js closures and global state on a current node.


So the focus it to extend `taskLib.submitNamed` API only.

**Comment by rymsha** [comment]
Initial idea was simple: disallow tasks to be run on specific nodes with `tasks = true/false` flag
After some probes it appeared that this is not enouth.

Consider this scenario:
- Cluster has nodes N1 and N2.
- N1 and N2 have `tasks = true` configuration.
- Local application A only installed on N1.
- A calls `taskLib.submitNamed(offload = true)`
- task gets offloaded to N2, but there is no A on it, and task fails.

Unfortunately same thing may easily happen with global applications as they are not guaranteed to be started when another node offloads a task.

We need applications to register themselves in the cluster so wen task gets submitted it gets delivered only to the node that can accept it.

**Comment by rymsha** [comment]
One of the goals is to allow only set of nodes to execute expensive computational tasks.
But I have a dilemma. Sometimes applications may submit a task earlier than any remote node is ready to accept it. So how XP should deal with it?
- execute task locally (even if task execution is not desired on this local node)
- throw an error (How easy is it from JS to handle exceptions, BTW?) <- this option I have already implemented
- wait in a loop retrying until submitting succeeds (configurable retries and timeouts of course )
- don't start applications before offloading is even possible (Hard to implement )

Any other options?

N.B.
Local task executor will always exist. It is necessary because some tasks are not "offloadable" - and must be executed locally anyway.

**Comment by rymsha** [comment]
I realized that to wait some little time does not make things worse. There is already a natural timeout due to nature of "distributed tasks" - they have to communicate via network. 

In a clustered setup applications wait for Hazelcast to start anyway: there is still a chance for application to submit a task before clustered task executor is ready (due to how OSGi works), but time frame is very little.
